sql/python/sqlflow/template/db.py

```diff
@@ -0,0 +1,177 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import abc
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class DataBase(object):
+    """A database interface """
+
+    def __init__(self, database, user, password, host, port):
+        self._database = database
+        self._user = user
+        self._password = password
+        self._host = host
+        self._port = port
+
+        self._db = self._connect()
+
+    def close(self):
+        self._db.close()
+
+    def execute(self, query):
+        """ Execute the query
+        Args:
+            query(str): the query to execute
+
+        Returns:
+            cursor(cursor):
+        """
+        cursor = self._db.cursor()
+        cursor.execute(query)
+        return cursor
+
+    def execute_many(self, query, values):
+        """ Execute the query with values
+        Args:
+            query(str): the query to execute
+            values(list(tuple)): the values
+
+        Returns:
+            cursor(cursor)
+        """
+        cursor = self._db.cursor()
+        cursor.executemany(query, values)
+        return cursor
+
+    def create_table(self, table_name, table_schema):
+        """ Create table
+        Args:
+            table_name(str): name of table
+            table_schema(list(tuple)): schema of table
+
+        Returns:
+            success if no error thrown out
+        """
+        statement = self._gen_create_table_statement(table_name, table_schema)
+        self.execute(statement)
+
+    def insert_values(self, table_name, table_schema, values):
+        """ Insert values to table
+        Args:
+            table_name(str): name of table
+            table_schema(list(str)): schema of table
+            values(list(tuple)): the value to insert
+
+        Returns:
+            success if no error thrown out
+        """
+        statement = self._gen_insert_table_statement(table_name, table_schema)
+        self.execute_many(statement, values)
+
+    def query_select(self, statement):
+        """ Select fields from table simply
+        Args:
+            statement(str): select statement
+
+        Return:
+            Returns:
+            field_names(list(str)): the names of fields
+            field_data(list): return data of column-style
+        """
+        cursor = self.execute(statement)
+        return self._retrieve_fields_and_value_from_cursor(cursor)
+
+    def _connect(self):
+        raise NotImplementedError("DataBase.connect")
+
+    def _gen_create_table_statement(self, table_name, table_schema):
```

Comment:
Which part of the submitter needs to create a table?

Reply:
> @uuleon I have no idea why things need to be implemented in such an overwhelming way, where
> 
> 1. Metaprogramming: `@six.add_metaclass(abs.ABSMeta)`
> 2. OOP : `classSQLDataBase(DataBase)`
> 3. ORM: `create_table`
> 
> are involved. Not to mention many methods are not implemented...
> 
> The goal of extracting python snippet from code template is to make the code more readable and testable. Please keep it simple... Readability matters...
> 
> I would suggest starting from simply taking out the code from the template like what @wangkuiyi did.
> 
> https://github.com/sql-machine-learning/sqlflow/blob/652aaaa4096f2bc2cb580ed7aa39ad81148c959d/sql/python/db.py#L1-L16

With all due respect, I don' think it's an overwhelming design. The goal of designing `DataBase` is to hide all the specific details between databases, to make the code template of the `submitter` clean and simple.

Here are two differences between databases:
1. get `field_names` from `cursor`
Hive:  [i[0][i[0].find('.')+1:] for i in cursor.description]
Others: [i[0] for i in cursor.description]

2. template of the insert statement
MySQL: insert into table_name (features, labels) values (%s, %s)
SQLite3: insert into table_name (features, labels) values(?, ?)
Hive: insert into table table_name (featuers, labels) values (%s, %s)

That's why class `DataBase` is created and OOP is involved. If OOP is removed, some methods still have to exist for resolve the conflictions mentioned above and the code is not simple as you wish, or maybe leave it in the code template of the submitter? I don't think it's a good idea.

Should we have to use `create_table` method before saving prediction result? Correct me if I am wrong, please.

At last,  only one method which is `_gen_insert_table_statement` in `HiveDataBase` is not implemented, not many as you mentioned.

Reply:
> At last, only one method which is _gen_insert_table_statement in HiveDataBase is not implemented, not many as you mentioned.

@uuleon I apologize for the statement "Not to mention many methods are not implemented". It is inaccurate. I didn't realize most `NotImplementedError` lives in the base class.

Reply:
> Should we have to use `create_table` method before saving prediction result? Correct me if I am wrong, please.

In the current implementation, we don't. Table creation is done at Go, so the Python program doesn't need to create_table. A few month ago @wangkuiyi and I discussed keeping the database complication in Go and leaving the Python program small.

> Here are two differences between databases: ...

You are right. I feel (2) is relatively easy to solve in Go by using different statements for different SQL engines. However, we haven't figured out how to solve (1) in Go. @weiguoz has a detailed description of (1) at https://github.com/sql-machine-learning/sqlflow/issues/318.